Repeater / templating / tests cleanup

fkleuver opened the pull request.

AureliaEffect commented:
This pull request **fixes 1 alert** when merging 7016dffb1e0a83068a2c19e4cdb3544d014e9169 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-12fc33a823149e7faea3e5684333df66b4a36620)

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 8d869f1e77ce4da3d51bdd46aec4f566d35a85f9 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-dd8cee9bbeb2789d2f11b7585f346e37fb461351)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging a5f9c9da290ab0bdf5134a058aff7725850e7b1f into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-77677209d5264f6b318ecc945c6b22edc8d085f3)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging b713da641f9846a2667cb5cacdc334bedc52ccd8 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-0ac883262e0b92316cd087984de3903fa9174c8e)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging d54d9d9b4a18990d8a7605ab46a9ecaa67c1a141 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-d6c64e83b36ab70ae47194a22c51f12a52c9367e)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bigopon commented:
Nice, if there are more passes coming for the repeater with array, I think we should add the items mutation integration as well? I'd like to see things like
```js
vm.items = [];
vm.items.push(newItem);
```
Maybe add a todo for that, together with some value converters

fkleuver commented:
@bigopon here are items mutation tests: https://github.com/aurelia/aurelia/pull/203/files#diff-52a7d44c6aeeb590d99b9fcb628d3718R178
There's an add (push), remove (pop) and move (reverse/sort). I could add more combinations and sequences but the loop count would explode, and that's kind of testing the array observer which already has every possible angle tested.

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 3def2ed8a471ef9e083b6b8cb1c1361863cf94ef into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-e95bf9618d0058761c65ac4dc8c0ef053181533e)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bigopon commented:
Yes, though the results are tested in different flushes. I was checking if it should be test with instance replacement + instance mutation combination in the same flush

fkleuver commented:
You got a point. I added some more, do you think all angles are covered or are we still missing some stuff?

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 2c432cf79a98bdbc38e02b7df0f653776ec42ac9 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-142d351fa2a483322d1274e9ca20171fa0d031b3)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bigopon commented:
@fkleuver Yes. That will boost the confidence and will make sure we don't start with code coverage of Repeat 👍

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 2c145b69bae7599957dba0519f9947145c744bdb into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-de860a27629de6cc475f41f4e084ca7bb1071ac2)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

fkleuver commented:
@EisenbergEffect @bigopon Just a side note here, after messing around for a bit I managed to get some basic integration tests up with the repeater + if/else, without using the compiler or the renderer. Although kind of tedious, this exercise did help me gain a better understanding of how templating works.
We can definitely eventually have some reusable test doubles to make testing complex interactions easier, both for ourselves and for end-users. I also see some potential for performance improvements by adding a specialized `INodeSequence` implementation for text bindings, and some other small perf optimizations via specialization. This can be a nice job for AOT.

**The tests**
If/Else appear to be a *little* bit quirky in terms of responsiveness (either that or I'm just misunderstanding how the lifecycles work). For example in the repeater+if/else tests, instance or item mutations don't seem to be picked up properly. Will need to look into this later. I'm calling it a day for now and I think this PR is OK for merging as it is now

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging ae29cc190599796b1e186fb05bcd755fb819533d into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-d22176a50361bdf70660e3f8c89105b761df2f93)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

EisenbergEffect commented on packages/runtime/src/templating/create-element.ts:
```diff
@@ -36,7 +36,10 @@ export class PotentialRenderable {
       templateOrNode: this.node,
       cache: 0,
-      build: {
+      build: typeof this.node === 'string' ? {
```
In what scenario would this come up?

EisenbergEffect commented on packages/runtime/src/templating/resources/repeat.ts:
```diff
@@ -1,269 +1,173 @@
-    this.local = this.sourceExpression.declaration.evaluate(flags, scope, null);
-    if (this.hasPendingInstanceMutation) {
-      this.changeSet.add(this);
+  public attaching(encapsulationSource: INode, lifecycle: IAttachLifecycleController): void {
```
The controller shouldn't be used directly in this case. The type should be `IAttachLifecycle` and you'll want to call `view.attach` instead. The controller interface should only be used by the actor that begins the lifecycle.

fkleuver replied:
Good to know. Is the mounting correct though?

EisenbergEffect replied:
That looks correct to me.

EisenbergEffect commented on packages/runtime/src/templating/resources/repeat.ts:
```diff
@@ -1,269 +1,173 @@
+    const { views } = this;
+    for (let i = 0, ii = views.length; i < ii; ++i) {
+      const view = views[i];
+      view.release();
```
if you call `release`, make sure you don't retain any references to the view after that because the runtime may return that view to the cache, in witch case you can't use it anymore.

EisenbergEffect replied:
Same note as above for the controller interface. This is circumventing optimizations.

fkleuver replied:
Makes sense, thanks for clarifying

EisenbergEffect replied:
I think we should find another name for `release` since it's really *allowing* the view to return to the cache whenever it is detached. In the code above, `release` needs to be called before `$detach` which is clearly un-intuitive based on the current name. So, we need to re-order these calls and then probably come up with a better name. Thoughts?

EisenbergEffect replied:
Actually, ignore that. I apparently code it to work correctly exactly as you used it ;)

EisenbergEffect replied:
Actually, double strike! 🤣 I was right the first time. We need to re-order these calls and then probably rename the API. Sorry. Sheesh.

fkleuver replied:
I tried both ways (I also tried clearing the views array) but this was the only combination that doesn't break almost all repeater tests. For some reason the nodes don't actually get removed with a different ordering. We need specific tests here I guess. I keep running into issues with this :P

fkleuver replied:
We need a bit clearer definitions on what the responsibilites of the lifecycles are, and I think you got the right idea with flags like `UnbindAfterDetached`. I think that's just evidence of that we may need to work towards a linkedlist that can handle full bind/attach/mount (and reverse) lifecycles. The other thing we need to watch out for is discrepancies in 1) call all detaches and then all unbinds for all components, or 2) call detach followed directly by unbind per component

EisenbergEffect commented on packages/runtime/test/unit/templating/create-element.spec.ts:
```diff
@@ -0,0 +1,175 @@
+import { eachCartesianJoinFactory, eachCartesianJoin, createElement, _ } from '../util';
+import { TargetedInstruction, INode, ICustomElementType, CustomElementResource, TargetedInstructionType } from '../../../src';
+
+describe(`createElement() creates element based on tag`, () => {
```
Awesome! Thanks for kickstarting the tests on this!

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 446d6fa1d3cd664ee053f4bf85c8cd75426baf77 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-c56b8b2c54d245012e19e2407c54ff379cb095c7)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

EisenbergEffect commented:
Shall we go ahead and merge this?

fkleuver commented:
I'm adding some lifecycle tests in a bit

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging a9bbc317bd66164a24ce0c2765f72d8446d365c9 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-d59bd35ad9afb7d31219c4718283b8df8d4b5c12)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 69ef253b6bc8741274f0b3d548ca4b4a7eb1b88b into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-6d41134452e21246527598a9a5b52f626ef8f9ca)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

fkleuver commented:
@EisenbergEffect okay I added a couple more tests, fixed a small bug with `AggregateLifecycleTask` (shadowed variable, and added a tslint rule to warn about that now), and added some specific errors to renderContext providers. Good to merge if you agree with the minor changes.

EisenbergEffect commented:
As much as I want to get this in for the nightly, I've actually got to run right now and won't be able to review it in time. I do think I can either review later tonight or first thing for me tomorrow morning at the latest. I'm looking forward to having a look. Thanks!

fkleuver commented:
Sure thing. I'll just add a couple more tests >:)

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging b8e6282e390d331ead6882b5d6be39131e859da2 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-61949f9ad9dbf66df6bc7489d7b25bb1d2de5c2e)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 70f85b5e2cbddc2fb29888943f1999e48351b9f7 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-22b1ed3e76559eb4c6c9387646ce1f0944f6c042)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging e2a22e61be9caba45a5b883a3dd30b09c3064d57 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-43191e3d35814deaaca83e8c9ec5775dbb4b70aa)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

fkleuver commented:
@EisenbergEffect added decorator and view tests, and a small fix in view.mount(), a check to throw a more meaningful error and an extra check to prevent an error with `noViewTemplate`. The view tests are a bit rough. I'd need to clean them up a bit and add better assertions later on (especially with child views), but for now they'll do the trick.

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 9c2cb8314c747a81fa6df9fd59e0269bef801260 into 5971d36d4b8629352b3d1ec29eb06cfddd0ef9c7 - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-3d8aebe9808a8445fce9fd364b0a8a76cad8e26d)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

bigopon commented:
Very nice base setup 👍
